### PR TITLE
initial playwright setup and test

### PR DESCRIPTION
**Pull request by Irev-Dev**
Resolves to #1038

Because clicking around in our app will often only change pixel in the stream and not HTML, it kind of breaks how these e2e tests usually work. I've explained our workaround in a few places, but probably the best explanation because it also has a bit of a tutorial on how to get started writing a test for our app is here (15min watch)

https://github.com/KittyCAD/modeling-app/assets/29681384/6f5e8e85-1003-4fd9-be7f-f36ce833942d

But the nuts and bolts of how we wait for changes from the engine is the react component in `src/components/EngineCommands.tsx`, but some changes to `src/lang/std/engineConnection.ts` were needed to.

Other random notes:
- We run on two different runners and hopefully will unify to just Ubuntu at some point, [more explanation here](https://github.com/KittyCAD/modeling-app/pull/1039/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R264)
- `data-` attributes is a sanctioned way to add arbitrary attributes to HTML, and they give us something easy to query for in the command log/debug panel. Note that `data-testid` is a common convention for testing frameworks, having said that if you had the element `<button data-testid="my-el">My Button</button>` the following query `page.getByRole('button', { name: 'My Button' })` is better than `await page.getByTestId('my-el')` because the former more closely represents how the user identities and interacts with the button. If the text within the button changed a lot because it uses a variable then using the testid might be warranted.

### CODE REVIEW

File: playwright.config.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * Read environment variables from file.
+ * https://github.com/motdotla/dotenv
+ */
+// require('dotenv').config();
+
+/**
+ * See https://playwright.dev/docs/test-configuration.
+ */
+export default defineConfig({
+  testDir: './src/e2e-tests',
```
Comment (pierremtb) [inline playwright.config.ts:13]: @Irev-Dev Do you think we could do something like `e2e/playwright`?

I put the tauri stuff at `e2e/tauri`, was thinking it made sense to (a) have them outside src* and (b) have both mechanisms in a single `e2e` top dir so we could get a quick understanding of coverage by looking at the list of spec files

Comment (Irev-Dev) [reply playwright.config.ts:13]: Yeah that's fair, I'll shuffle things.

Comment (Irev-Dev) [reply playwright.config.ts:13]: Thanks for thinking these things through 👏

File: e2e/playwright/test-utils.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
+import { expect, Page } from '@playwright/test'
+import { EngineCommand } from '../../src/lang/std/engineConnection'
+
+async function waitForPageLoad(page: Page) {
+  // wait for 'Loading KittyCAD Modeling App...' spinner
+  await page.waitForFunction(() =>
+    document.querySelector('[data-testid="initial-load"]')
```
Comment (pierremtb) [inline e2e/playwright/test-utils.ts:7]: TIL about `data-testid` 😃

File: e2e/playwright/flow-tests.spec.ts [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+document.addEventListener('mousemove', (e) =>
+  console.log(`await page.mouse.click(${e.clientX}, ${e.clientY})`)
+)
+*/
+
+test.beforeEach(async ({ context, page }) => {
+  context.addInitScript(async (token) => {
+    localStorage.setItem('TOKEN_PERSIST_KEY', token)
+    localStorage.setItem('persistCode', ``)
```
Comment (pierremtb) [inline e2e/playwright/flow-tests.spec.ts:20]: @Irev-Dev alright so this is how you handle auth in web tests? Cool that it's a more direct mechanism compared to #1040

File: e2e/playwright/test-utils.ts (hunk 2 above)
Comment (Irev-Dev) [reply e2e/playwright/test-utils.ts:7]: Oh nice, yeah bit of a convention. Were you familiar with `data-` attributes in general?

File: src/components/Stream.tsx [hunk 4]
```diff
@@ -356,6 +356,8 @@ export const Stream = ({ className = '' }) => {
 
         kclManager.executeAstMock(modifiedAst, true)
       })
+    } else {
+      engineCommandManager.sendSceneCommand(command)
```
Comment (Irev-Dev) [inline src/components/Stream.tsx:360]: We were never sending mouse_drag_end events to the engine, but it didn't seem to be a problem 🤷

**Comment by Irev-Dev** [comment]
Alright, I think this one's ready for a review @franknoirot @pierremtb.

Since there's possibly a lot of new stuff here for all of us, maybe pay special attention to the readme updates. lmk if anything doesn't make sense or there's something missing.

**Comment by Irev-Dev** [comment]
[8b0915c](https://github.com/KittyCAD/modeling-app/pull/1039/commits/8b0915c3c94695e2919b0d758f79c3c3a0e00f6f) failed for reasons unrelated to playwright.

**Comment by mxschmitt** [comment]
Looks very nice! Saw this PR on Twitter and wanted to give some feedback:

I saw that there are a bunch of awaits missing in `setViewportSize` /  `addInitScript` / `emulateMedia`. We recommend the following ESLint rule '@typescript-eslint/no-floating-promise' to prevent that as per [here](https://playwright.dev/docs/best-practices#lint-your-tests) to prevent missing awaits which will/could lead to unexpected behaviour, since you don't ensure that the api call was completed.

Also instead of doing `waitForFunction` with `querySelector` we recommend `waitFor()` on the locator, e.g. `page.getByRole(...).waitFor()`. This will give better error reporting in trace-viewer.

I saw you are doing `toMatchSnapshot` for screenshots, if you use `toHaveScreenshot` instead, they will retry and wait for two consecutive page screenshots yield the same result etc. which is more robust.

nit: `getUtils` could be converted into a fixture, so that you can use it in other tests as well instead of having to call it all the time.

Its also possible to run an individual test by line number, e.g. `yarn playwright test foo:23` (file which contains foo). Have you tried [UI Mode](https://playwright.dev/docs/test-ui-mode)?

All these are minor issues, looks great!

**Comment by Irev-Dev** [comment]
Awesome @mxschmitt!

Heaps of good tips there thank you!

I'll get cracking on those and no I hadn't tried UI mode. I'll give that a go.

File: e2e/playwright/test-utils.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+async function expectCmdLog(page: Page, locatorStr: string) {
+  await expect(page.locator(locatorStr)).toBeVisible()
+}
+
+async function waitForDefaultPlanesToBeVisible(page: Page) {
+  await page.waitForFunction(
+    () =>
+      document.querySelectorAll('[data-receive-command-type="object_visible"]')
+        .length >= 3
+  )
```
Comment (Irev-Dev) [inline e2e/playwright/test-utils.ts:43]: I couldn't figure out how to make a locator that waited for three of the same selector to appear, so kept this for the time being.

**Review by Irev-Dev — COMMENTED** [review]
Thanks again @mxschmitt, and of course thanks for your work on such a great tool!

I integrated all of your suggestions in [5f4cec1](https://github.com/KittyCAD/modeling-app/pull/1039/commits/5f4cec15024e23fe065a77d03883066b501c4ea2) with the exception of fixtures, I want to think about that a little more but created #1116.

I also have one remaining `waitForFunction` as noted below.

Also, UI mode is great, one minor inconvenience is the video only shows up in the timeline, but only minor.

image

But one great thing about UI mode is that it works with `Google Chrome`, as the "record new" and "record at cursor" extension features seemed to only launch with chromium, and we have a video encoding compat issue with that, so being able to use locator feature in UI-mode is great.

**Comment by Irev-Dev** [comment]
I'm going to merge this, even though the playwright CI is failing, it's because our dev engine has an issue right now. CI was passing before and the tests work locally.

**Pull request merged**